src/vm/gcheaputilities.h

```diff
@@ -192,11 +195,96 @@ class GCHeapUtilities {
     }
 #endif // FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
 
-
 private:
     // This class should never be instantiated.
     GCHeapUtilities() = delete;
 };
 
+// When not using per-thread allocation contexts, we (the EE) need to take care that
+// no two threads are concurrently modifying the global allocation context. This lock
+// must be acquired before any sort of operations involving the global allocation context
+// can occur.
+//
+// This lock is acquired by all allocations when not using per-thread allocation contexts. 
+// It is acquired in two kinds of places:
+//   1) JIT_TrialAllocFastSP (and related assembly alloc helpers), which attempt to 
+//      acquire it but move into an alloc slow path if acquiring fails
+//      (but does not decrement the lock variable when doing so)
+//   2) Alloc and AllocAlign8 in gchelpers.cpp, which acquire the lock using
+//      the Acquire and Release methods below.
+class GlobalAllocLock {
```

Comment:
I tried to keep all of the GC utility functions and structs in one place with `gcheaputilities.h`, but I'd be fine moving it to gchelpers as well. It seems to me that gchelpers is more for allocation helpers - nobody including `gchelpers.h` (and only that file) would need to use `GlobalAllocLock`, while includers of `gcheaputilities.h` would (since they may be calling methods on IGCHeap directly).